cuda/matrix/dense_kernels.cu

```diff
@@ -428,27 +679,199 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
+{
+    extern __shared__ size_type block_max[];
+
+    reduce_array(
+        size, nnz_per_row, block_max,
+        [](const size_type &x, const size_type &y) { return max(x, y); });
+
+    if (threadIdx.x == 0) {
+        result[blockIdx.x] = block_max[0];
+    }
+}
+
+
+}  // namespace kernel
+
+
 template <typename ValueType>
 void calculate_max_nnz_per_row(std::shared_ptr<const CudaExecutor> exec,
                                const matrix::Dense<ValueType> *source,
-                               size_type *result) NOT_IMPLEMENTED;
+                               size_type *result)
+{
+    const auto num_rows = source->get_size()[0];
+    auto nnz_per_row = Array<size_type>(exec, num_rows);
+
+    calculate_nonzeros_per_row(exec, source, &nnz_per_row);
+
+    const auto n = ceildiv(num_rows, default_block_size);
+    const size_type grid_dim =
+        (n <= default_block_size) ? n : default_block_size;
+
+    auto block_results = Array<size_type>(exec, grid_dim);
+
+    kernel::reduce_max_nnz<<<grid_dim, default_block_size,
+                             default_block_size * sizeof(size_type)>>>(
+        num_rows, as_cuda_type(nnz_per_row.get_const_data()),
+        as_cuda_type(block_results.get_data()));
+
+    auto d_result = Array<size_type>(exec, 1);
+
+    kernel::reduce_max_nnz<<<1, default_block_size,
+                             default_block_size * sizeof(size_type)>>>(
+        grid_dim, as_cuda_type(block_results.get_const_data()),
+        as_cuda_type(d_result.get_data()));
+
+    exec->get_master()->copy_from(exec.get(), 1, d_result.get_const_data(),
+                                  result);
+    d_result.clear();
+    block_results.clear();
+    nnz_per_row.clear();
+}
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(
     GKO_DECLARE_DENSE_CALCULATE_MAX_NNZ_PER_ROW_KERNEL);
 
 
+namespace kernel {
+
+
+template <typename ValueType>
+__global__ __launch_bounds__(default_block_size) void count_nnz_per_row(
+    size_type num_rows, size_type num_cols, size_type stride,
+    const ValueType *__restrict__ work, size_type *__restrict__ result)
+{
+    __shared__ size_type part_results[default_block_size];
+    const auto warp_size = cuda_config::warp_size;
+    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+    const auto global_warpid = tidx / warp_size;
+    const auto local_warpid = threadIdx.x / warp_size;
+
+    if (global_warpid < num_rows) {
+        part_results[threadIdx.x] = 0;
+        for (auto i = threadIdx.x % warp_size; i < num_cols; i += warp_size) {
+            if (work[stride * global_warpid + i] != zero<ValueType>()) {
+                part_results[threadIdx.x] += 1;
+            }
+        }
+        __syncthreads();
+        if (threadIdx.x % warp_size == 0) {
+            result[global_warpid] = 0;
+            for (auto i = 0; i < warp_size; i++) {
+                result[global_warpid] +=
+                    part_results[local_warpid * warp_size + i];
+            }
+        }
```

Comment:
Comming back to that after I tried something, but this is not an efficient way of doing this.

Reply:
You can use the following function that uses `reduce`, which, in turn, uses efficient shuffle operations which work inside a warp.

You need to add `#include "cuda/components/cooperative_groups.cuh"` to your includes and change the function to this:
```c++
template <typename ValueType>
__global__ __launch_bounds__(default_block_size) void count_nnz_per_row(
    size_type num_rows, size_type num_cols, size_type stride,
    const ValueType *__restrict__ work, size_type *__restrict__ result)
{
    constexpr auto warp_size = cuda_config::warp_size;
    const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
    const auto row_idx = tidx / warp_size;

    if (row_idx < num_rows) {
        size_type part_result{};
        for (auto i = threadIdx.x % warp_size; i < num_cols; i += warp_size) {
            if (work[stride * row_idx + i] != zero<ValueType>()) {
                part_result += 1;
            }
        }
        
        auto warp_tile = group::tiled_partition<warp_size>(group::this_thread_block());
        result[row_idx] = reduce(warp_tile, part_result, [](const size_type &a, const size_type &b) {return a + b;});
    }
}
```

Reply:
With this the compilation for cuda 9.0 fails. What do you suggest, should I revert this to my latest working version?

Reply:
@fritzgoebel can you add a `__syncthreads();` before the `auto warp_tile = [...]` and see if that works?

Reply:
CUDA 9.1 was released in December 2017. I don't mind dropping support for versions older than that. However, I think it should work with 9.0...
https://devblogs.nvidia.com/cooperative-groups/

Reply:
There was missing as_cuda_type in the kernel call, I hope it works now.

If it does: my bad, sorry!

Reply:
> @fritzgoebel can you add a `__syncthreads();` before the `auto warp_tile = [...]` and see if that works?

Please don't use CUDA's raw synchronization and communication functions. Use cooperative groups instead. You know what happened with shuffles....

Reply:
In general, all intra-thread operations (synchronization, communication, memory fences, votes, etc.) should go through the cooperative groups API. This way, we have a single place where we can update things by adding additional data, or call alternative intrinsics if NVIDIA decides to change something.

If something is missing from cooperative groups, rather add it to the API [here](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/components/cooperative_groups.cuh) then using the intrinsic directly in code.

Maybe this is worth writing down somewhere, like the contributing guide or something...